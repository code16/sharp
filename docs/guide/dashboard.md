# Create a Dashboard

A Dashboard is a good way to present synthetic data to the user, with graphs, stats, or personalized reminders for instance.

## Generator

```bash
php artisan sharp:make:dashboard <class_name>
```

## Write the class

A Dashboard is very much like an Entity Form, except it is readonly. So the first step is to create a new class extending `Code16\Sharp\Dashboard\SharpDashboard` which leads us to implement three functions:

- `buildWidgets()`, similar to Entity Form's `buildForm()`
- `buildWidgetsLayout()`, similar to `buildLayout()`
- `buildDashboardConfig()`, for optional filters
- and `buildWidgetsData(DashboardQueryParams $params)`, for the actual Dashboard data, like Entity Form's `find()` method.

### `buildWidgets()`

We're supposed to use here `$this->addWidget()` to configure all the Dashboard widgets.

```php
function buildWidgets()
{
    $this->addWidget(
        SharpLineGraphWidget::make("capacities")
            ->setTitle("Spaceships by capacity")

    )->addWidget(
        SharpPanelWidget::make("activeSpaceships")
            ->setInlineTemplate("<h1>{{count}}</h1> spaceships in activity")
            ->setLink('spaceship')
    );
}
```

As we can see in this example, we defined two widgets giving them a mandatory `key` and some optional properties depending of their type.

Every widget has the optional following setters:

- `setTitle(string $title)` for the widget title displayed above it
- `setLink(string $entityKey, string $instanceId = null, array $querystring = [])` to make the whole widget linked to a specific entity. To link to the Entity List, pass the `$entityKey`, and add the `$instanceId` to link to the Entity Form.

And here's the full list and documentation of each widget available, for the specifics:

- [Graph](dashboard-widgets/graph.md)
- [Panel](dashboard-widgets/panel.md)
- [OrderedList](dashboard-widgets/ordered-list.md)

### `buildWidgetsLayout()`

The layout API is a bit different of Entity Form here, because we think in terms of rows and not columns. So for instance:

```php
function buildWidgetsLayout()
{
    $this->addFullWidthWidget("capacities")
        ->addRow(function(DashboardLayoutRow $row) {
            $row->addWidget(6, "activeSpaceships")
                ->addWidget(6, "inactiveSpaceships");
        });
}
```

We can only add rows and "full width widgets" (which are a shortcut for a single widget row). A row groups widgets in a 12-based grid.

### `buildWidgetsData(DashboardQueryParams $params)`

Widget data is set with specific methods depending of their type. The documentation is therefore split:

- [Graph](dashboard-widgets/graph.md)
- [Panel](dashboard-widgets/panel.md)
- [OrderedList](dashboard-widgets/ordered-list.md)

## Configure the Dashboard

Once this class written, we have to declare the form in the sharp config file:

```php
// config/sharp.php

return [
    "entities" => [
        [...]
    ],
    "dashboards" => [
        "company_dashboard" => [
            "view" => \App\Sharp\CompanyDashboard::class
        ]
    ],
    [...]
    "menu" => [
        [
            "label" => "Company",
            "entities" => [
                [
                    "label" => "Dashboard",
                    "icon" => "fa-dashboard",
                    "dashboard" => "company_dashboard"
                ],
                [...]
            ]
        ]
    ]
];
```

In the menu, like an Entity, a Dashboard can be displayed anywhere.

## Dashboard filters

Just like EntityLists, Dashboard can display filters, as [documented on the Filter page](filters.md).

## Dashboard commands

Like again EntityLists, Commands can be attached to a Dashboard: [see the Command documentation](commands.md).

## Dashboard policies

Just like for an Entity, you can define a Policy for a Dashboard. The only available action is `view`.

```php
// config/sharp.php

return [
    "entities" => [
        [...]
    ],
    "dashboards" => [
        "company_dashboard" => [
            "view" => \App\Sharp\CompanyDashboard::class,
            "policy" => \App\Sharp\Policies\CompanyDashboardPolicy::class,
        ]
    ],
    [...]
];
```

And the policy class can be pretty straightforward:

```php
class CompanyDashboardPolicy
{
    public function view(User $user)
    {
        return $user->hasGroup("boss");
    }
}
```

