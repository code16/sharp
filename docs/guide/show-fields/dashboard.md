# Dashboard

Class: `Code16\Sharp\Show\Fields\SharpShowDashboardField`.

The field allows you to integrate a [Dashboard](../building-dashboard.md) into your Show Page.

## Constructor

This field needs, as first parameter, either the entity key or the `SharpDashboardEntity` class that declares the dashboard which will be included in the Show Page.

For instance:

```php
SharpShowDashboardField::make('posts_dashboard')
```

or

```php
SharpShowDashboardField::make(PostDashboardEntity::class)
```
::: warning
This last syntax is better in terms of DX (since it allows using the IDE to navigate to the Entity List implementation), but it won’t work in two specific cases: if you use a custom `SharpEntityResolver` or if you your Entity is declared with multiple keys.
:::

## Configuration

### `hideFilterWithValue(string $filterName, $value)`
This is the most important method of the field, since it will not only hide a filter but also set its value. The purpose is to allow to **scope the data to the instance** of the Show Page. For example, let’s say we display a Post and that we want to embed a dashboard with the post's statistics:


```php
class PostShow extends SharpShow
{
    // ...
    
    public function buildShowFields(FieldsContainer $showFields): void
    {
        $showFields->addField(
            SharpShowDashboardField::make(PostDashboardEntity::class)
                ->hideFilterWithValue(PostFilter::class, 64)
        );
    }
}
```

Here we're scoping the `PostDashboard` declared in the `PostDashboardEntity` to the instance of the `Post` with id 64. 


You can pass a closure as the value, and it will contain the current Show instance id. In most cases, you'll have to write this:

```php
SharpShowDashboardField::make(PostDashboardEntity::class)
    ->hideFilterWithValue(PostFilter::class, fn ($instanceId) => $instanceId);
```

**One final note**: sometimes the linked filter is really just a scope, never displayed to the user. In this case, it can be tedious to write a full implementation in the Dashboard. In this situation, you can use the `HiddenFiler` class for the filter, passing a key:

```php
class PostShow extends SharpShow
{
    // ...
    
    public function buildShowFields(FieldsContainer $showFields): void
    {
        $showFields->addField(
            SharpShowDashboardField::make(PostDashboardEntity::class)
                ->hideFilterWithValue('post', fn ($instanceId) => $instanceId);
        );
    }
}
```

```php
use \Code16\Sharp\EntityList\Filters\HiddenFilter;

class PostDashboard extends SharpDashboard
{
    // ...

    protected function getFilters(): ?array
    {
        return [
            HiddenFilter::make('post')
        ];
    }
    
    protected function buildWidgetsData(): void
    {
        return $this->setFigureData('visit_count', 
            figure: Post::query()
                ->findOrFail($this->queryParams->filterFor('post'))
                ->get()?->visit_count
        );
    }
}
```

### `hideDashboardCommand(array|string $commands): self`

Use it to hide any dashboard command in this particular Dashboard (useful when reusing a Dashboard). This will apply before looking at authorizations.

## Display in layout

To display your dashboard in your show page's layout, you have to use the `addDashboardSection()` method in your Show Page's `buildShowLayout()` method.

```php
protected function buildShowLayout(ShowLayout $showLayout): void
    {
        $showLayout
            ->addSection(function (ShowLayoutSection $section) {
                $section
                    ->addColumn(7, function (ShowLayoutColumn $column) {
                        $column
                            ->withFields(categories: 5, author: 7)
                           // ...
                    })
                    ->addColumn(5, function (ShowLayoutColumn $column) {
                        // ...
                    });
            })
            ->addDashboardSection(PostDashboardEntity::class);
    }
```
