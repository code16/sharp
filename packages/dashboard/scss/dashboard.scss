
@import 'components/Widget';
@import 'components/Chart';
@import 'components/Panel';
@import 'components/Legend';
@import 'components/OrderedList';
@import "page";