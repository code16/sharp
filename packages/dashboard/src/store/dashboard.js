import {
    getDashboard,
    postDashboardCommand,
    getDashboardCommandFormData,
} from "../api";
import { filtersModule as filters } from 'sharp-filters';
import { commandsModule as commands } from 'sharp-commands';

export const UPDATE = 'UPDATE';
export const SET_DASHBOARD_KEY = 'SET_DASHBOARD_KEY';

export default {
    namespaced: true,
    modules: {
        filters,
        commands,
    },
    state: {
        dashboardKey: null,
        data: null,
        widgets: null,
        config: null,
        layout: null,
    },
    mutations: {
        [UPDATE](state, { data, layout, widgets, config }) {
            state.data = data;
            state.widgets = widgets;
            state.layout = layout;
            state.config = config;
        },
        [SET_DASHBOARD_KEY](state, dashboardKey) {
            state.dashboardKey = dashboardKey;
        }
    },
    actions: {
        update({ commit, dispatch }, { data, widgets, layout, config, filtersValues }) {
            commit(UPDATE, {
                data,
                widgets,
                layout,
                config,
            });
            return Promise.all([
                dispatch('filters/update', {
                    filters: config.filters,
                    values: filtersValues
                }),
                dispatch('commands/update', {
                    commands: config.commands
                })
            ]);
        },
        async get({ state, dispatch, getters }, { filtersValues }) {
            const data = await getDashboard({
                dashboardKey: state.dashboardKey,
                filters: getters['filters/getQueryParams'](filtersValues)
            });
            await dispatch('update', {
                data: data.data,
                widgets: data.widgets,
                layout: data.layout,
                config: data.config,
                filtersValues,
            });
        },
        postCommand({ state }, { command, data, query }) {
            return postDashboardCommand({
                dashboardKey: state.dashboardKey,
                commandKey: command.key,
                data,
                query,
            });
        },
        getCommandFormData({ state }, { command, query }) {
            return getDashboardCommandFormData({
                dashboardKey: state.dashboardKey,
                commandKey: command.key,
                query,
            });
        },
        setDashboardKey({ commit }, dashboardKey) {
            commit(SET_DASHBOARD_KEY, dashboardKey);
        }
    }
}