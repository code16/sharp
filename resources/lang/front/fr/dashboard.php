<?php

return [
    'commands.dashboard.label' => 'Actions',
];
