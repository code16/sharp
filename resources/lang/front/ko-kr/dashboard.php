<?php

return [
    'commands.dashboard.label' => '동작',
];
