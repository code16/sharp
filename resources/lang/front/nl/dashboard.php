<?php

return [
    'commands.dashboard.label' => 'Acties',
];
