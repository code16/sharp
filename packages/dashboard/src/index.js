import dashboard from './store/dashboard';
import routes from './routes';

export default function(Vue, { router, store }) {
    store.registerModule('dashboard', dashboard);
    router.addRoutes(routes);
}