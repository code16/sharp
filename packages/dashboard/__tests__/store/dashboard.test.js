import dashboard, { UPDATE} from '../../src/store/dashboard';

describe('store dashboard', ()=>{
    test('state matches snapshot', ()=>{
        expect(dashboard.state).toMatchSnapshot();
    });

    describe('mutations', ()=>{
        test('UPDATE', ()=>{
            const state = {};
            const data = {};
            const widgets = {};
            const layout = {};
            const config = {};
            dashboard.mutations[UPDATE](state, {
                data, widgets, layout, config
            });
            expect(state.data).toBe(data);
            expect(state.widgets).toBe(widgets);
            expect(state.layout).toBe(layout);
            expect(state.config).toBe(config);
        });
    });
});