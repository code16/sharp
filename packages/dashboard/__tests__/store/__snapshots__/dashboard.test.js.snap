// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store dashboard state matches snapshot 1`] = `
Object {
  "config": null,
  "dashboardKey": null,
  "data": null,
  "layout": null,
  "widgets": null,
}
`;
